src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -152,121 +91,199 @@ def __init__(
+                FlowExecutor.create,
+                flow_file=self._flow_file,
+                connections=self._connections,
+                working_dir=self._working_dir,
+                raise_ex=False,
+            )
+
+            # 1. Create input_queue, output_queue, and _process_info in the main process.
+            # 2. Spawn _n_process sub-process and pass the above queue/dict to these sub-process to transfer information
+            # between main process and sub process.
+            self._processes_manager = SpawnProcessManager(
+                executor_creation_func,
+                self._input_queues,
+                self._output_queues,
+                self._process_info,
+                _process_wrapper,
+                False,
+            )
+
+        self._processes_manager.start_processes()
+
+        monitor_pool = ThreadPool(self._n_process, initializer=set_context, initargs=(contextvars.copy_context(),))
+        self._monitor_pool = monitor_pool
 
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        if self._pool is not None:
-            self._pool.close()
-            self._pool.join()
+        if self._monitor_pool is not None:
+            self._monitor_pool.close()
+            self._monitor_pool.join()
+
+    def _get_process_info(self, index):
+        while True:
+            try:
+                process_id = self._process_info[index].process_id
+                process_name = self._process_info[index].process_name
+                return (index, process_id, process_name)
+            except KeyError:
+                continue
+            except Exception as e:
+                bulk_logger.warning(f"Unexpected error occurred while get process info. Exception: {e}")
+
+    def handle_line_timeout(self, line_number, timeout_time, inputs, run_id, start_time, result_list):
+        bulk_logger.warning(f"Line {line_number} timeout after {timeout_time} seconds.")
+        ex = LineExecutionTimeoutError(line_number, timeout_time)
+        result = self._generate_line_result_for_exception(inputs, run_id, line_number, self._flow_id, start_time, ex)
+        result_list.append(result)
 
-    def _timeout_process_wrapper(self, run_start_time: datetime, task_queue: Queue, timeout_time, result_list):
-        healthy_ensured_process = HealthyEnsuredProcess(self._executor_creation_func, self.context)
-        healthy_ensured_process.start_new(task_queue)
+    def handle_process_crashed(self, line_number, inputs, run_id, start_time, result_list):
+        bulk_logger.warning(f"Process crashed while executing line {line_number},")
+        ex = ProcessCrashError(line_number)
+        result = self._generate_line_result_for_exception(inputs, run_id, line_number, self._flow_id, start_time, ex)
+        result_list.append(result)
 
-        if not healthy_ensured_process.process.is_alive():
-            return
+    def _monitor_process_alive(self, process_id):
```

Comment:
`is_process_alive` might be more straightforward?

Reply:
fixed